control-plane/api-gateway/controllers/reference_validator.go

```diff
@@ -0,0 +1,174 @@
+package controllers
+
+import (
+	"context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+type ReferenceValidator struct {
+	client.Client
+}
+
+func NewReferenceValidator(client client.Client) *ReferenceValidator {
+	return &ReferenceValidator{
+		client,
+	}
+}
+
+func (rv ReferenceValidator) GatewayCanReferenceSecret(ctx context.Context, gateway gwv1beta1.Gateway, secretRef gwv1beta1.SecretObjectReference) (bool, error) {
```

Comment:
It will be slightly more efficient to make these pointer receivers to avoid copying the client on every function call.